geode-core/src/main/java/org/apache/geode/distributed/internal/membership/gms/locator/GMSLocator.java

```diff
@@ -70,52 +71,54 @@
   private final String securityUDPDHAlgo;
   private final String locatorString;
   private final List<HostAddress> locators;
+  private final LocatorStats locatorStats;
+  private final Set<InternalDistributedMember> registrants = new HashSet<>();
+  private final Map<InternalDistributedMemberWrapper, byte[]> publicKeys =
+      new ConcurrentHashMap<>();
+
+  private volatile boolean isCoordinator;
+
   private Services services;
-  private final LocatorStats stats;
   private InternalDistributedMember localAddress;
 
-  private final Set<InternalDistributedMember> registrants = new HashSet<>();
-  private Map<InternalDistributedMemberWrapper, byte[]> registerMbrVsPK = new ConcurrentHashMap<>();
-
   /**
    * The current membership view, or one recovered from disk. This is a copy-on-write variable.
    */
-  private transient NetView view;
+  private NetView view;
 
-  private transient NetView recoveredView;
+  private NetView recoveredView;
 
   private File viewFile;
 
-  private volatile boolean isCoordinator;
-
   /**
    * @param bindAddress network address that TcpServer will bind to
    * @param locatorString location of other locators (bootstrapping, failover)
    * @param usePreferredCoordinators true if the membership coordinator should be a Locator
    * @param networkPartitionDetectionEnabled true if network partition detection is enabled
-   * @param stats the locator statistics object
+   * @param locatorStats the locator statistics object
    * @param securityUDPDHAlgo DF algorithm
    */
   public GMSLocator(InetAddress bindAddress, String locatorString, boolean usePreferredCoordinators,
-      boolean networkPartitionDetectionEnabled, LocatorStats stats, String securityUDPDHAlgo) {
+      boolean networkPartitionDetectionEnabled, LocatorStats locatorStats,
+      String securityUDPDHAlgo) {
     this.usePreferredCoordinators = usePreferredCoordinators;
     this.networkPartitionDetectionEnabled = networkPartitionDetectionEnabled;
     this.securityUDPDHAlgo = securityUDPDHAlgo;
     this.locatorString = locatorString;
-    if (this.locatorString == null || this.locatorString.length() == 0) {
-      this.locators = new ArrayList<>(0);
+    if (this.locatorString == null || this.locatorString.isEmpty()) {
+      locators = new ArrayList<>(0);
     } else {
-      this.locators = GMSUtil.parseLocators(locatorString, bindAddress);
+      locators = GMSUtil.parseLocators(locatorString, bindAddress);
     }
-    this.stats = stats;
+    this.locatorStats = locatorStats;
   }
 
   @Override
   public synchronized boolean setMembershipManager(MembershipManager mgr) {
     if (services == null || services.isStopped()) {
       services = ((GMSMembershipManager) mgr).getServices();
       localAddress = services.getMessenger().getMemberID();
-      assert localAddress != null : "member address should have been established";
+      Objects.requireNonNull(localAddress, "member address should have been established");
```

Comment:
So `assert` requires the `-enableassertions` flag to be set in order to actually have checks run.  `assert` also throws `AssertionError` unlike other validity checks.  `requireNonNull` was introduced in Java 7 to allow better null checking with a message.  See Effective Java, 3rd edition for more details.  If this statement was only here for debugging, then we could delete it, but it is likely we don't want to attempt to process at all if the value is `null` and throw an exception to be handled.

Reply:
Michael explained it best (and with a reference!). In general, we should avoid using `assert` statements. If the check is worth keeping have having enabled all of the time, then it should be converted to a check that's always enabled (hence my change to requireNonNull).

Reply:
Thanks for the explanation. I think `assert` is still useful when you specifically want a condition checked while debugging and want it to be a no-op in production (think hot paths). I also think Effective Java item 49 does not say to avoid using `assert` in favor of `requireNonNull`, but rather to use `requireNonNull` instead of checking for null manually, e.g. `if (someVariable == null) throw new NullPointerException();`.